### PR TITLE
[Adjustments] Admin adjustments

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Makes some final changes to "admin" adjustments and clears up a lot of remaining adjustments mess.

Also updates some bits of UI and behavior in the order adjustments tab:
- Shipping and payment fees can't be arbitrarily changed. They're calculated automatically based on the relevant shipping or payment methods. The option to edit them is removed.
- Display of taxes in the adjustments tab is updated to show more info
- When creating an adjustment manually, the tax category is selected instead of the tax rate (the corresponding tax rate is applied automatically based on the category).
- Correctly "adds" the tax amount when creating new manual adjustments that use **added tax** (for US and CA). this is a change in behavior that needs some announcement to users. The `amount` entered will be without tax, and the tax will then be added on top. Previously users had to try to figure out what the after-tax total would be, and calculate backwards to figure out what to put as the amount, which was really annoying and counter-intuitive.

Review note: the red lines in these diffs are more important than the green ones :wink:

#### What should we test?

Adding and updating manual adjustments via the order edit adjustments tab.

#### Release notes

Updated admin adjustments

Changelog Category: User facing changes

### CODE REVIEW

File: app/services/order_tax_adjustments_fetcher.rb [hunk 1]
```diff
@@ -10,25 +9,13 @@ def initialize(order)
 
   def totals
     order.all_adjustments.tax.each_with_object({}) do |adjustment, hash|
-      tax_rates_hash = tax_rates_hash(adjustment)
-      hash.update(tax_rates_hash) { |_tax_rate, amount1, amount2| amount1 + amount2 }
+      tax_rate = adjustment.originator
+      tax_amounts = { tax_rate => adjustment.amount }
+      hash.update(tax_amounts) { |_tax_rate, amount1, amount2| amount1 + amount2 }
```
Comment (mkllnk) [inline app/services/order_tax_adjustments_fetcher.rb:14]: Wow, so before, each adjustment could have multiple tax rates but now it's only one, right?

We could probably simplify this further:
```suggestion
      tax_rate = adjustment.originator
      hash[tax_rate] = hash[tax_rate].to_f + adjustment.amount
```

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
+class MigrateAdminTaxAmounts < ActiveRecord::Migration[6.0]
+  class Spree::Adjustment < ApplicationRecord
+    belongs_to :originator, polymorphic: true
```
Comment (mkllnk) [inline db/migrate/20210617203927_migrate_admin_tax_amounts.rb:3]: I don't see this relationship used. Would it be easier to omit this line and see `originator_id` and `originator_type` as simple attributes?

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    belongs_to :adjustable, polymorphic: true
+    belongs_to :order, class_name: "Spree::Order"
```
Comment (mkllnk) [inline db/migrate/20210617203927_migrate_admin_tax_amounts.rb:5]: I don't see this one used either. We don't actually want to load `Spree::Order`. We just use the `order_id` column which doesn't need to know about the Order code.

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    belongs_to :tax_category, class_name: 'Spree::TaxCategory'
```
Comment (mkllnk) [inline db/migrate/20210617203927_migrate_admin_tax_amounts.rb:6]: And here I'm wondering as well if we could just use `tax_category_id` instead.

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    has_many :adjustments, as: :adjustable, dependent: :destroy
```
Comment (mkllnk) [inline db/migrate/20210617203927_migrate_admin_tax_amounts.rb:7]: I don't see this used in the migration.

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    scope :admin, -> { where(originator_type: nil) }
+  end
+
+  def up
+    migrate_admin_taxes!
+  end
+
+  def migrate_admin_taxes!
+    Spree::Adjustment.admin.where('included_tax <> 0').find_each do |adjustment|
```
Comment (mkllnk) [inline db/migrate/20210617203927_migrate_admin_tax_amounts.rb]: Is it correct to assume that none of these adjustments have the tax adjustment already? For example, is all the current production code setting `included_tax` to `0`?

File: spec/migrations/migrate_admin_tax_amounts_spec.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  let(:adjustment50) { create(:adjustment, amount: 100, included_tax: 50) }
+
+  describe '#migrate_admin_taxes!' do
+    context "when the adjustment has no tax" do
+      let!(:adjustment_without_tax) { create(:adjustment, included_tax: 0) }
+
+      it "doesn't move the tax to an adjustment" do
+        expect(Spree::Adjustment).to_not receive(:create!)
+
+        subject.migrate_admin_taxes!
```
Comment (mkllnk) [inline spec/migrations/migrate_admin_tax_amounts_spec.rb:23]: This requires knowledge of the internal working (using `create!`) which may change in the future (eg `new(..).save!`. I've seen this solution in the past:
```suggestion
        expect { subject.migrate_admin_taxes! }.to_not change {
          Spree::Adjustment.count
        }
```

**Review by mkllnk — APPROVED** [review]
Yah! Some really good clean up.

I'm approving this pending all my questions don't raise issues. :wink:

File: app/views/spree/admin/adjustments/_adjustments_table.html.haml [hunk 8]
```diff
@@ -19,5 +19,6 @@
         %td.align-center.included-tax= adjustment.display_included_tax.to_html
         - unless @order.canceled?
           %td.actions
-            = link_to_edit adjustment, no_text: true
-            = link_to_delete adjustment, no_text: true
+            - if adjustment.originator_type.nil?
+              = link_to_edit adjustment, no_text: true
+              = link_to_delete adjustment, no_text: true
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/views/spree/admin/adjustments/_adjustments_table.html.haml:34]: There are already some issues around this in master where the current logic doesn't always work well or produces unexpected results. If the user manually edits the amount on a shipping or payment fee adjustment, the value will be reset when they hit "save and recalculate fees", as those values are calculated based on the related shipping or payment methods.

So here we're removing those options to avoid confusion. So in that scenario you outlined, the hub manager could change the shipping method to one with a different cost. Otherwise yes, manual adjustments for negative or positive values can be added at any time, so if the hub manager wanted to refund $5 on shipping because a friend picked up the order that day, they could just add an adjustment for `-$5`, and label it `"Partial shipping refund for pickup"` (or whatever they like).

How does that sound?

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb (hunk 6 above)
Comment (Matt-Yorkley) [reply db/migrate/20210617203927_migrate_admin_tax_amounts.rb]: "Admin" adjustments are the only remaining type (currently) that still make use of the `included_tax` attribute and don't have tax adjustments.

After this, we won't be using that database column any more for any adjustment types, all tax amounts will be recorded in tax adjustments. We can even drop that column from the table, but I thought I'd wait.

File: app/services/order_tax_adjustments_fetcher.rb (hunk 1 above)
Comment (Matt-Yorkley) [reply app/services/order_tax_adjustments_fetcher.rb:14]: > Wow, so before, each adjustment could have multiple tax rates but now it's only one, right?

Multiple tax adjustments for different rates can still be added to a single item! :heavy_check_mark: 

In this case though, we're looking at tax adjustments that have _already been created_ (`order.all_adjustments.tax`) and we're summing them by rate.

The complexity in this class was related to the previous use of the `included_tax` attribute, where theoretically we could have been squashing multiple tax amounts from multiple rates into a single database column (with no reference to the rates). Figuring out what rates were used in that case meant we had to attempt to reverse-engineer it to see if multiple rates had been used, and then try to compute both. Now that all tax amounts are stored in _distinct adjustments_ (which each clearly reference the rate that was used to create them), that problem totally disappears.

Does that make sense? :sweat_smile:

File: app/services/tax_rate_finder.rb [hunk 9]
```diff
@@ -48,37 +45,4 @@ def line_item_tax_category(enterprise_fee, line_item)
...
-  # - If an admin created an adjustment with a miscalculated included tax then
-  #   we don't know which tax rate the admin intended to use.
-  # - An admin may also enter included tax that doesn't correspond to any tax
-  #   rate in the system. They may enter a fee of $1.2 with tax of $0.2, but
-  #   that doesn't mean that there is a 20% tax rate in the database.
-  # - The used tax rate may also have been deleted. Maybe the tax law changed.
-  #
-  # In either of these cases, we will find a tax rate that doesn't correspond
-  # to the included tax.
-  def find_closest_tax_rates_from_included_tax(amount, included_tax)
```
Comment (Matt-Yorkley) [inline app/services/tax_rate_finder.rb:78]: I don't think I've ever been so happy about deleting code! It's been a wild journey getting to the point where this is possible :fire::fire::fire:

File: app/views/spree/admin/adjustments/_adjustments_table.html.haml (hunk 8 above)
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/views/spree/admin/adjustments/_adjustments_table.html.haml:34]: Yes, sounds much cleaner and clearer for everyone.

File: app/services/order_tax_adjustments_fetcher.rb (hunk 1 above)
Comment (mkllnk) [reply app/services/order_tax_adjustments_fetcher.rb]: Yes. And it makes me really appreciate the great work you've done to clean up this mess. :medal_sports:

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb (hunk 6 above)
Comment (mkllnk) [reply db/migrate/20210617203927_migrate_admin_tax_amounts.rb]: So if we were super thorough, we would deploy the UI changes first and then run the migration so that nobody can set included tax while we are running it. That scenario seems very unlikely though.

**Review by mkllnk — APPROVED** [review]

File: db/migrate/20210617203927_migrate_admin_tax_amounts.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply db/migrate/20210617203927_migrate_admin_tax_amounts.rb:3]: I changed this migration a couple of times whilst working on it. I'll see if it can be tidied up a bit :ok_hand:

File: app/controllers/spree/admin/adjustments_controller.rb [hunk 10]
```diff
@@ -43,58 +41,13 @@ def skip_changing_canceled_orders
...
-      # closed. But open adjustments cannot be edited.
-      # To preserve updates, like changing the amount of the shipping fee,
-      # we close the adjustment first.
-      #
-      # The Spree admin interface allows to open and close adjustments manually
-      # but we removed that functionality as it had no purpose for us.
-      def enable_updates
-        @adjustment.close
+      def apply_tax
+        Spree::TaxRate.adjust(@order, [@adjustment])
```
Comment (andrewpbrett) [inline app/controllers/spree/admin/adjustments_controller.rb:45]: 💯

**Review by andrewpbrett — APPROVED** [review]
Amazing work @Matt-Yorkley 🎉 It's a lot of changes but they're easy to understand. The point by Maikel about the migrations would be nice to clean up but apart from that this looks outstanding. Moving to Test Ready.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

I had a look at the key points you mentioned, in a added-tax server (staging-UK):

**1)**
> Shipping and payment fees can't be arbitrarily changed; The option to edit them is removed.

Before - options to edit/delete are visible:

image

After - options to edit/delete are gone!

image


**2)** 
> Display of taxes in the adjustments tab is updated to show more info

In the "after"-pic above we can see, that the `Tax Category` value for Shipping is blank. However, this shipping method does have tax category set:

image

i) So, it looks like this is not getting fetched to the Adjustments table
ii) Also, the `Tax` value - in this case, 0.26 (2x0.13) - is not displayed on the table.

**3)**
> When creating an adjustment manually, the tax category is selected instead of the tax rate

This is great, and works perfectly:
image


**4)** 
> Correctly "adds" the tax amount when creating new manual adjustments that use added tax (for US and CA). 

Previously users had to try to figure out what the after-tax total would be, and calculate backwards to figure out what to put as the amount, which was really annoying and counter-intuitive.


**5)**

I just noticed that `Shipping` does not appear in the Order Adjustments section

image

I guess all is good :tada:
The exceptions are:
- 2.i) and 2.ii) - issue #7932, which was considering the previous table
- 5) for which I would open if we choose to tackle this separately

I think this PR already adds considerable improvement - points 1) 2) 3) - so I'll check staging-FR (inclusive tax) for these changes and if all is well we could merge this.

**Comment by filipefurtad0** [comment]
Hum, just spotted an issue with printing invoices, `/admin/orders/<order_nr>/print`:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/60f04c0d02ce690007efa56f?event_id=60f04c0d007f8e9a84180000&i=sk&m=nw

This happens for alternative invoices - regular invoices get printed as expected (but I guess are not very used around added tax instances).

I'll move to In Dev for now :+1:

**Comment by filipefurtad0** [comment]
The test cases above all yield the same result. On test case 3) the adjustments tab indicates that the tax is inclusive :+1: 
image

Calculates the included tax amount and adds it to the order total correctly.

**Tl;dr,** issues to address:
2.i) and 2.ii) Displaying Tax Category and Tax for the Shipping fee
5) Including Shipping fee in the Order adjustments section
6) Alternative invoice printing is broken

**Comment by Matt-Yorkley** [comment]
Okay, I think it's all fixed up and ready for re-testing those last couple of display issues. Regarding the order adjustments section:
> 5) Including Shipping fee in the Order adjustments section
I think it hasn't been shown there for a while, but it id shown higher up the page under the shipment bit.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Great! I've gone through the previous tests and missing points.

**Added-tax instances**

Points 1) 3) and 4) remain addressed by this PR

2.i) and 2.ii) We can now see that the Shipping fee appears correctly under the Adjustments tab:

image

The `Tax Category`, `Tax Amount` and the `Total` appear correctly :+1: This closes #7932.

6) The alternative invoice now renders :+1: 

However, for the **regular invoice** we should take notice that the amount for the `Sales Tax` column now appears displayed for added adjustments, but is displaying a zero value for Shipping. Before this PR, the sales tax column was empty (before, left; after, right):

image

I think this might be a bit misleading.

**Inclusive-tax instances** - points 1) 3) and 4) remain addressed by this PR


Points 2.i) and 2.ii) are addressed as well:

image

The `Tax Category`, `Tax Amount` and the `Total` appear correctly :+1: This closes #7932.

6) Invoices render as before, but we may have a similar issue, in the GST column, for the **alternative invoice**:

image

The regular invoice is displayed as before:

image

So, we're good to merge this?

On issues:
5) Yeah, indeed - this was already the case before this PR. I'm not sure this is necessary since the correct tax amount now appears under `Adjustments`, also for new added adjustments. Is it a requirement for the `Tax` amounts for Shipping and added adjustments be displayed on the order details tab as well?

6) These improvements on invoices are perhaps best dealt with on a separate issue?

**Comment by Matt-Yorkley** [comment]
Thanks! I'll take a quick look at that last invoice bit and maybe add a couple of extra specs here as regression tests.
